Backport of:

From 07244a83014fad42da937c17d98474b47a570bf7 Mon Sep 17 00:00:00 2001
From: "Gregory P. Smith" <greg@krypto.org>
Date: Wed, 24 May 2017 00:04:38 -0700
Subject: [PATCH] bpo-9146: Raise a ValueError if OpenSSL fails to init a hash
 func. (#1777)

This helps people in weird FIPS mode environments where common things
like MD5 are not available in the binary as a matter of policy.
---
 Modules/_hashopenssl.c | 45 +++++++++++++++++++++++++++++++++---------
 1 file changed, 36 insertions(+), 9 deletions(-)

Index: python3.5-3.5.2/Modules/_hashopenssl.c
===================================================================
--- python3.5-3.5.2.orig/Modules/_hashopenssl.c	2019-07-10 07:57:12.781176599 -0400
+++ python3.5-3.5.2/Modules/_hashopenssl.c	2019-07-10 07:57:50.461312006 -0400
@@ -86,7 +86,10 @@ EVP_hash(EVPobject *self, const void *vp
             process = MUNCH_SIZE;
         else
             process = Py_SAFE_DOWNCAST(len, Py_ssize_t, unsigned int);
-        EVP_DigestUpdate(&self->ctx, (const void*)cp, process);
+        if (!EVP_DigestUpdate(&self->ctx, (const void*)cp, process)) {
+            PyErr_SetString(PyExc_ValueError, "unsupported hash type");
+            break;
+        }
         len -= process;
         cp += process;
     }
@@ -143,7 +146,10 @@ EVP_digest(EVPobject *self, PyObject *un
 
     locked_EVP_MD_CTX_copy(&temp_ctx, self);
     digest_size = EVP_MD_CTX_size(&temp_ctx);
-    EVP_DigestFinal(&temp_ctx, digest, NULL);
+    if (!EVP_DigestFinal(&temp_ctx, digest, NULL)) {
+        PyErr_SetString(PyExc_ValueError, "unsupported hash type");
+        return NULL;
+    }
 
     retval = PyBytes_FromStringAndSize((const char *)digest, digest_size);
     EVP_MD_CTX_cleanup(&temp_ctx);
@@ -163,7 +169,10 @@ EVP_hexdigest(EVPobject *self, PyObject
     /* Get the raw (binary) digest value */
     locked_EVP_MD_CTX_copy(&temp_ctx, self);
     digest_size = EVP_MD_CTX_size(&temp_ctx);
-    EVP_DigestFinal(&temp_ctx, digest, NULL);
+    if (!EVP_DigestFinal(&temp_ctx, digest, NULL)) {
+        PyErr_SetString(PyExc_ValueError, "unsupported hash type");
+        return NULL;
+    }
 
     EVP_MD_CTX_cleanup(&temp_ctx);
 
@@ -288,7 +297,12 @@ EVP_tp_init(EVPobject *self, PyObject *a
             PyBuffer_Release(&view);
         return -1;
     }
-    EVP_DigestInit(&self->ctx, digest);
+    if (!EVP_DigestInit(&self->ctx, digest)) {
+        PyErr_SetString(PyExc_ValueError, "unsupported hash type");
+        if (data_obj)
+            PyBuffer_Release(&view);
+        return -1;
+    }
 
     self->name = name_obj;
     Py_INCREF(self->name);
@@ -387,7 +401,11 @@ EVPnew(PyObject *name_obj,
     if (initial_ctx) {
         EVP_MD_CTX_copy(&self->ctx, initial_ctx);
     } else {
-        EVP_DigestInit(&self->ctx, digest);
+        if (!EVP_DigestInit(&self->ctx, digest)) {
+            PyErr_SetString(PyExc_ValueError, "unsupported hash type");
+            Py_DECREF(self);
+            return NULL;
+        }
     }
 
     if (cp && len) {
@@ -727,6 +745,8 @@ generate_hash_name_list(void)
  *  the generic one passing it a python string and are noticably
  *  faster than calling a python new() wrapper.  Thats important for
  *  code that wants to make hashes of a bunch of small strings.
+ *  The first call will lazy-initialize, which reports an exception
+ *  if initialization fails.
  */
 #define GEN_CONSTRUCTOR(NAME)  \
     static PyObject * \
@@ -743,6 +763,17 @@ generate_hash_name_list(void)
         if (data_obj) \
             GET_BUFFER_VIEW_OR_ERROUT(data_obj, &view); \
      \
+        if (CONST_new_ ## NAME ## _ctx_p == NULL) { \
+            EVP_MD_CTX *ctx_p = &CONST_new_ ## NAME ## _ctx; \
+            if (!EVP_get_digestbyname(#NAME) || \
+                !EVP_DigestInit(ctx_p, EVP_get_digestbyname(#NAME))) { \
+                _setException(PyExc_ValueError); \
+                EVP_MD_CTX_cleanup(ctx_p); \
+                return NULL; \
+            } \
+            CONST_new_ ## NAME ## _ctx_p = ctx_p; \
+        } \
+     \
         ret_obj = EVPnew( \
                     CONST_ ## NAME ## _name_obj, \
                     NULL, \
@@ -767,10 +798,6 @@ generate_hash_name_list(void)
 #define INIT_CONSTRUCTOR_CONSTANTS(NAME)  do { \
     if (CONST_ ## NAME ## _name_obj == NULL) { \
         CONST_ ## NAME ## _name_obj = PyUnicode_FromString(#NAME); \
-        if (EVP_get_digestbyname(#NAME)) { \
-            CONST_new_ ## NAME ## _ctx_p = &CONST_new_ ## NAME ## _ctx; \
-            EVP_DigestInit(CONST_new_ ## NAME ## _ctx_p, EVP_get_digestbyname(#NAME)); \
-        } \
     } \
 } while (0);
 
